### PR TITLE
Try to bind just one function

### PR DESCRIPTION
**Pull request by solidsnack**
This one is entertaining. In it, we declare and use a single function from `pgcommon`. Linking with `crate build` fails. However, I take the command it prints out as having failed and add `-l pgcommon` and then `target/debug/elephantpump` works!

```
"cc" "-m64" "-L" "/usr/local/Cellar/rust/1.3.0/lib/rustlib/x86_64-apple-darwin/lib" "/Users/solidsnack/sandbox/elephantpump/target/debug/elephantpump.0.o" "-o" "/Users/solidsnack/sandbox/elephantpump/target/debug/elephantpump" "-Wl,-force_load,/usr/local/Cellar/rust/1.3.0/lib/rustlib/x86_64-apple-darwin/lib/libmorestack.a" "-Wl,-dead_strip" "-nodefaultlibs" "/Users/solidsnack/sandbox/elephantpump/target/debug/deps/liblibc-540159808ccfa9ab.rlib" "/usr/local/Cellar/rust/1.3.0/lib/rustlib/x86_64-apple-darwin/lib/libstd-198068b3.rlib" "/usr/local/Cellar/rust/1.3.0/lib/rustlib/x86_64-apple-darwin/lib/libcollections-198068b3.rlib" "/usr/local/Cellar/rust/1.3.0/lib/rustlib/x86_64-apple-darwin/lib/librustc_unicode-198068b3.rlib" "/usr/local/Cellar/rust/1.3.0/lib/rustlib/x86_64-apple-darwin/lib/librand-198068b3.rlib" "/usr/local/Cellar/rust/1.3.0/lib/rustlib/x86_64-apple-darwin/lib/liballoc-198068b3.rlib" "/usr/local/Cellar/rust/1.3.0/lib/rustlib/x86_64-apple-darwin/lib/liblibc-198068b3.rlib" "/usr/local/Cellar/rust/1.3.0/lib/rustlib/x86_64-apple-darwin/lib/libcore-198068b3.rlib" "-L" "/Users/solidsnack/sandbox/elephantpump/target/debug" "-L" "/Users/solidsnack/sandbox/elephantpump/target/debug/deps" "-L" "/usr/local/lib/" "-L" "/usr/local/lib/" "-L" "/usr/local/Cellar/rust/1.3.0/lib/rustlib/x86_64-apple-darwin/lib" "-L" "/Users/solidsnack/sandbox/elephantpump/.rust/lib/x86_64-apple-darwin" "-L" "/Users/solidsnack/sandbox/elephantpump/lib/x86_64-apple-darwin" "-l" "System" "-l" "pthread" "-l" "c" "-l" "m" "-l" "compiler-rt" -l pgcommon
```

The `build.rs` does not seem to do anything. Behold, the results of `pg_str_endswith()`:

```
:;  target/debug/elephantpump
A ends with A: 1
A ends with B: 0
```

### CODE REVIEW

**Comment by posix4e** [comment]
The sys library was for getting the struct types in postgresql over as well. Might we still need it for that?

**Comment by solidsnack** [comment]
We should definitely revisit the sys library. We might be able to do everything with just a handful of structs and functions, though.

No matter what route we take, we should have linking working in the context of `cargo build`. Not sure what's wrong with my `build.rs` that's cause the reference to `pgcommon` to be dropped.

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+extern crate libc;
+
+extern "C" {
```
Comment (solidsnack) [inline src/lib.rs]: It turns out that `-l pgcommon` is passed when the lib is built -- not when the binary is built.

Comment (posix4e) [inline src/lib.rs]: Yes
On Mon, Dec 21, 2015 at 9:57 PM Jason Dusek notifications@github.com
wrote:

> In src/lib.rs
> https://github.com/posix4e/elephantpump/pull/10#discussion_r48224040:
> 
> > @@ -0,0 +1,7 @@
> > +extern crate libc;
> > +
> > +extern "C" {
> 
> It turns out that -l pgcommon is passed when the lib is built -- not when
> the binary is built.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/posix4e/elephantpump/pull/10/files#r48224040.
> 
> ## 

Sent from my mobile 5045076749
Alex Newman

**Comment by posix4e** [comment]
Lgtm feel free to merge as long as travis merges

**Comment by solidsnack** [comment]
So, we are back to the giant file of bindings. Once I tried tracing the ones for logical decoding I realized that it's like 300 lines of stuff by itself; and very messy, so what's the point.

Right now I am using `allow(...)` as an annotation to silence warnings when that codes is built. If we go back to the old layout, I assume we can do the same thing from the `Cargo.toml` in `elephant-sys`?

There are some symbols missing in this binding, like `pg_str_endswith`. This is part of `pgcommon`. If you could share your technique for generating headers, maybe as a Makefile (?) then we could try extending it to cover really everything.

**Comment by posix4e** [comment]
I noticed you merged the giant file of bindings into the same package. Any reason not to have a separate sys bindings package? It seems to be the norm for rust.

**Comment by solidsnack** [comment]
My understanding is that a `*-sys` package is usually:
- Named for the underlying library (`libpq-sys` or `pq-sys` or maybe `pg-sys`),
- Offers comprehensive bindings,
- Declares `links` and there is some rule about only one package ever can `links` a native lib.

I anticipate in two or three weeks we'll be able to offer a real `pg-sys`. One thing we need to sort out is missing symbols, like those from `pgcommon`. We might also need two or three of these package - one for clients, one for servers. But in the near term it would seem to be making a promise we don't want to focus, namely that of a comprehensive binding to Postgres libraries. I am trying to stay focused on a JSON based streaming logical decoder at present.

**Comment by posix4e** [comment]
What still is needed to be done to merge this? Can we use my branch which generates the rustbindings dynamically?

**Comment by posix4e** [comment]
I'm going to let you pull in the autosys branch

**Pull request merged**